[No QA] Create React-Native Tooltip

roryabraham opened the pull request:
cc @marcaaron @tgolen 

### Details
This implements a native tooltip component in React Native, preventing us from needing to adopt an external web-only library such as React-Tooltip or material-ui. For now, tooltips will only work on web/desktop, because they are dependent on `Hoverable`.

For popovers, I think we can continue to use `react-native-modal`. If necessary, this component _could_ be abstracted to be a more generalizable popover component.

One known limitation is that tooltips are still subject to the stacking context of the component they're used on:

![image](https://user-images.githubusercontent.com/47436092/105890892-c8ade380-5fc4-11eb-8cec-119fddb71680.png)

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/147470

### Tests
Use the tooltip on various components like so:

```
<Tooltip text="Do something">
    <MyOtherComponent />
</Tooltip>
```

Then on web/desktop hover over the component you wrapped with a tooltip. Make sure the tooltip appears. Play around with it and look for any edge cases. I discovered a few while developing this, which I've done my best to address.

On iOS/Android/mWeb, just do some basic regression testing and make sure nothing is broken.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots

#### Web
![image](https://user-images.githubusercontent.com/47436092/105890710-913f3700-5fc4-11eb-8f6d-5ba465f7450b.png)
![image](https://user-images.githubusercontent.com/47436092/105890779-a320da00-5fc4-11eb-99fb-81a52dd29f60.png)

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
![image](https://user-images.githubusercontent.com/47436092/106046765-87d2cf00-6097-11eb-97bd-3d16f30025db.png)

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

roryabraham committed:
Create tooltip pointer

roryabraham committed:
Create Hoverable component

roryabraham committed:
Create first draft tooltip

roryabraham committed:
Fix animation and hover state

roryabraham committed:
Rename child ref to wrapperView

roryabraham committed:
Update tooltip style

roryabraham committed:
Save draft state - need horizontal positioning

roryabraham committed:
Get horizontal shifting working

roryabraham committed:
Fix js style

roryabraham committed:
Move all styles to getTooltipStyles and nix TooltipPointer component

roryabraham committed:
Make tooltip appear below content if near the top of the screen

roryabraham committed:
Implement horizontal shifting

shawnborton commented:
If we can have a tooltip like this:
![image](https://user-images.githubusercontent.com/2319350/105892822-aa6ed480-6012-11eb-9b2f-dc694e23d795.png)


Then does that mean we can avoid this situation where the tooltip runs over the left edge? Can we have this tooltip look left-aligned?
![image](https://user-images.githubusercontent.com/2319350/105892923-c3778580-6012-11eb-8baa-cc034e5e377f.png)


Also, minor design feedback, can we add just a little bit of horizontal padding to the tooltips?

shawnborton commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,112 @@
+        tooltipWrapperStyle: {
+            position: 'absolute',
+            backgroundColor: themeColors.heading,
+            borderRadius: 4,
```
We should have a global variable for this.

shawnborton replied:
Or rather, we could create one for this? Maybe like `componentBorderRadiusSmall: 4`

roryabraham committed:
Remove unnecessary state

roryabraham committed:
Improve comments in style

roryabraham committed:
Add method docs in tooltip

roryabraham committed:
Remove unnecessary directory

tgolen commented on src/components/Hoverable/index.js:
```diff
@@ -0,0 +1,57 @@
+        // Clone the child element, providing the callbacks to the react element to handle hover state changes.
+        // Using `React.Children.only` enforces that this component has only one child,
+        // which gives the `Hoverable` a more predictable UX.
+        return React.cloneElement(React.Children.only(child), {
```
Why use `cloneElement()` here (and what does it do)? Some more comment would be helpful since `cloneElement()` has never been used in our code before.

tgolen commented on src/components/Hoverable/index.native.js:
```diff
@@ -0,0 +1,15 @@
+// Simply render the children, ignoring all other `Hoverable` props.
+// Using React.Children.only enforces that only a single child element is provided,
+// which matches the desired interface of `Hoverable` on web/desktop.
+const Hoverable = props => React.Children.only(props.children);
```
Does this still work if `props.children` is a function (since it seems to matter in the web component)?

tgolen commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+    text: PropTypes.string.isRequired,
+
+    // Child to wrap with Tooltip. Must be only child.
+    children: PropTypes.oneOfType([
```
`PropTypes.node`

roryabraham committed:
Make variable for small border radius

tgolen commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    /**
+     * Get the position and measure the size of the wrapper view and the tooltip itself.
```
> Get the position

Clarify what this is getting the position of (it's not clear)

tgolen commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+    /**
+     * Get the position and measure the size of the wrapper view and the tooltip itself.
+     */
+    getPosition() {
```
This could maybe be renamed to `measureViews` or `measureWrapperAndTooltipViews` and then the purpose of the method might be more intuitive.

roryabraham commented:
> If we can have a tooltip like this...does that mean we can avoid this situation where the tooltip runs over the left edge?

We could! All that would be involved is adding some props to the tooltip like `shiftHorizonal` and `shiftVertical` and manually setting them when the tooltip needs to be shifted. The only reason I didn't do that is because I wasn't sure there would ever be a real need.

A much better solution in my opinion would be to have the tooltip show up over any other content in the window no matter what, and only shift if we need to avoid the edge of the window. I couldn't figure out how to make that work though, because the tooltip is subject to the stacking context it lives in, so `zIndex` doesn't solve the problem. I'm open to any ideas, or we could just live with the limitation and solve that problem later if we need to.

tgolen commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+    tooltipHeight,
+) {
+    // Determine if the tooltip should display below the wrapped component.
+    const windowWidth = Dimensions.get('window').width;
```
How is `Dimensions.get('window').width` different from `width` in the param?

roryabraham replied:
Hopefully this is made clear by the docs I added on this method.

tgolen commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+    return 0;
+}
+
+export default function getTooltipStyles(
```
docs

roryabraham commented:
Updated padding:

![image](https://user-images.githubusercontent.com/47436092/105906650-c5bcee00-5fd8-11eb-9226-c2e4f107c729.png)

roryabraham committed:
Update padding

tgolen reviewed (changes requested):
Very nice implementation here! I'm curious why `src/libs/getPopoverCoordinates.js` is included as an empty file?

It would be nice to see it actually implemented somewhere in this PR (ie there is no way to QA this), but I guess that's coming in another PR which is OK.

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+            >
+                <View
+                    ref={el => this.wrapperView = el}
+                    onLayout={this.getPosition}
```
Does `onLayout` return layout props we can use? https://reactnative.dev/docs/view#onlayout

roryabraham replied:
Sorry, not sure I understand your question...

From what I can tell, `onLayout` provides its callback 4 arguments:

1. `x` -> the distance between the left side of the window and the left side of the view
2. `y` -> the distance between the top of the window and the top of the view
3. `width` -> the width of the view
4. `height` -> the height of the view

marcaaron replied:
Right, so we then call `getPosition()` which seems to return the same stuff when it calls `measureInWindow()` ?

https://reactnative.dev/docs/direct-manipulation#measureinwindowcallback

So another way to phrase my question would be... why do we need `measureInWindow()` and can we just use the layout props returned by `onLayout`?

roryabraham replied:
Ahh, I remember why. `onLayout` is relative to the parent element, but `measureInWindow` gives you the position relative to the window, which is what we want in this case I think.

https://github.com/facebook/react-native/issues/10556

See how they're different:

![image](https://user-images.githubusercontent.com/47436092/105921375-9f557d80-5fed-11eb-83cf-60f7418e7ada.png)

roryabraham replied:
So my above comment wasn't correct. `onLayout` gives you:

1) `x` -> the distance between the left side of the parent and the left side of the view
2) `y` -> the distance between the top of the parent and the top of the view.

roryabraham replied:
Added a comment in the code explaining why we use `measureInWindow`

roryabraham committed:
Remove empty file

shawnborton commented:
What if you are on the desktop app and you make the app smaller than 800px
wide?

On Tue, Jan 26, 2021 at 10:11 PM Rory Abraham <notifications@github.com>
wrote:

> If we can have a tooltip like this...does that mean we can avoid this
> situation where the tooltip runs over the left edge?
>
> We could! All that would be involved is adding some props to the tooltip
> like shiftHorizonal and shiftVertical and manually setting them when the
> tooltip needs to be shifted. The only reason I didn't do that is because I
> wasn't sure there would ever be a real need.
>
> A much better solution in my opinion would be to have the tooltip show up
> over any other content in the window no matter what, and only shift if we
> need to avoid the edge of the window. I couldn't figure out how to make
> that work though, because the tooltip is subject to the stacking context it
> lives in, so zIndex doesn't solve the problem. I'm open to any ideas, or
> we could just live with the limitation and solve that problem later if we
> need to.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Expensify/Expensify.cash/pull/1335#issuecomment-767830403>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARWH5VTVY7XVLUW5H7NA73S34VY3ANCNFSM4WT6JGLQ>
> .
>

roryabraham commented:
> I'm curious why src/libs/getPopoverCoordinates.js is included as an empty file?

That's a mistake - removed!

> It would be nice to see it actually implemented somewhere in this PR (ie there is no way to QA this), but I guess that's coming in another PR which is OK.

Not sure if you think this is better, but I created a feature branch for the `ReportActionContextMenu` (formerly known as the `CommentActionsMenu`), so I could break up that one feature into multiple more reasonably-sized PRs. I could change the base branch of this PR to be that feature branch, so that it will be QA'd along with the `ReportActionContextMenu`. Could also just merge it to master 🤷🏼‍♂️ What do you think?

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+        const interpolatedSize = this.animation.interpolate({
+            inputRange: [0, 1],
+            outputRange: [0, 1],
+        });
```
Just wondering why do we need the `interpolate`? Not super familiar with what it is doing and was hoping you could help me understand.

roryabraham replied:
Yeah, I'll do my best to give a hand-wavy explain, but it's really only to the best of my understanding. I bet @Luke9389 could explain maybe better (given his background in teaching math, as this is really just linear algebra mumbo jumbo).

Basically `interpolate` changes a number gradually over the set duration of the animation. It defines how to map the inputs provided for an animation to the output result of the animation, over the duration of the animation.

You could provide different parameters for the input range and output range, but this maps `0 -> 0`, `0.25 -> 0.25`, etc... basically something like this:

- At time 0%, the interpolation factor will be 0
- At time 50%, the interpolation factor will be 0.5
- At time 100%, the interpolation factor will be 1

You can think of that as the "identity function" of interpolation. Whereas if we did this:

```
this.animation.interpolate({
    inputRange: [0, 1],
    outputRange: [0, 2],
});
```

We would end up with:

- At time 0%, the interpolation factor will be 0.
- At time 50%, the interpolation factor will be 1
- At time 100%, the interpolation factor will be 2.

`showTooltip` triggers the animation to begin with its starting value of 0 and end with a value of 1, so we end up with something like:

- At time 0%, the animation value of 0 is interpolated to 0 so the size is 0.
- At time 50%, the animation value of 0.5 is interpolated to 0.5, so the size is 0.5.
- At time 100%, the animation value of 1 is interpolated to 1, so the size is 1.

`hideTooltip` triggers the animation  to begin with its starting value of 1 and end with a value of 0, so we end up with something like:

- At time 0%, the animation value of 1 is interpolated to 1, so the size is 1.
- At time 50%, the animation value of 0.5 is interpolated to 0.5, so the size is 0.5.
- At time 100%, the animation value of 0 is interpolated to 0, so the size is 0.

I think that's mostly correct. Hopefully I didn't just confuse you more 😅

marcaaron replied:
Ok, can we just animate our value from `0` to `2`?

marcaaron replied:
And in this case we are mapping the input to the same output range?

Luke9389 replied:
Here's a [link to the docs](https://reactnative.dev/docs/animations#interpolation) on animation interpolation. I'm not sure that we need it for `showTooltip` (bc our input is the same as our output), but I think we would still need it for `hideTooltip`.

marcaaron replied:
Thanks for the link @Luke9389! Maybe I'm still missing it. Why would we need it for `hideTooltip`?

roryabraham replied:
@marcaaron Is there some concern you have about this implementation? What alternative would you suggest?

marcaaron replied:
Well I'm just trying to understand why this is here first... but my concern is that this seems unnecessary. As an alternative, I'd suggest to use the animated value directly instead of the interpolated value.

roryabraham replied:
You were right, the interpolation was unnecessary. It's removed now

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+                <View
+                    ref={el => this.wrapperView = el}
+                    onLayout={this.getPosition}
+                    collapsable={false}
```
What will happen if this is `collapsable={true}`?

roryabraham replied:
I'm not 100% sure if it's completely necessary, but from the docs:

> Views that are only used to layout their children or otherwise don't draw anything may be automatically removed from the native hierarchy as an optimization. Set [collapsable] to false to disable this optimization and ensure that this View exists in the native view hierarchy.

I was just trying to make sure that `this.wrapperView.measureInWindow` doesn't error because this wrapper view doesn't render anything that's not in another view.

marcaaron replied:
Gooot it. Sorta sounds like the docs are saying that if this has no visible children it will be removed from the view hierarchy. But this makes sense to me since if this `View` had no visible children then it should also not need a `Tooltip` either? 🤔  

Anyways, I'm still not sure what we are preventing. Maybe we can add this when it becomes a solution to a problem we've run into? Unless you've already observed that problem then we should add a comment.

roryabraham replied:
Yeah, my concern was that the tooltip itself would not always be visible, so it would be removed from the native hierarchy and not measured before it needs to be shown. I'll try remove it and see if it causes any problems

tgolen commented:
I think it's OK to merge this to `master` and just have it tested in other PRs.

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+    tooltipHeight,
+) {
+    // Determine if the tooltip should display below the wrapped component.
+    const windowWidth = Dimensions.get('window').width;
```
We maybe should pass this into this method as an optimization. We should only need to know if the window width has updated when it changes. Recently added an HOC we can use for this here.

https://github.com/Expensify/Expensify.cash/blob/master/src/components/withWindowDimensions.js

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+import fontFamily from './fontFamily';
+import variables from './variables';
+
+const GUTTER_WIDTH = 16;
```
**Idea for a comment:** Explain what is the gutter width used for and why is it `16`

marcaaron replied:
Should it be a constant in `variables`?

roryabraham replied:
I'm not sure `GUTTER_WIDTH` will be relevant anywhere else, so I think it should be left here. I'll add a comment tho

roryabraham committed:
Fix prop types

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+) {
+    // Determine if the tooltip should display below the wrapped component.
+    const windowWidth = Dimensions.get('window').width;
+    const shouldShowBelow = (yOffset - tooltipHeight) < GUTTER_WIDTH;
```
**Idea for a comment:** Explain what we are measuring here, why it should influence `shouldShowBelow`, and what the effect of setting `shouldShowBelow` to true or false will be.

roryabraham committed:
Add jsDoc for getTooltipStyles

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+    xOffset,
+    yOffset,
+    width,
+    height,
```
`width` and `height` should have more descriptive names? It's not clear what they relate to.

roryabraham committed:
Simplify callback functions used in Hoverable

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+    // from displaying too near to the edge of the screen.
+    let horizontalShift = 0;
+    const tooltipLeftEdge = (xOffset + (width / 2)) - (tooltipWidth / 2);
+    const tooltipRightEdge = (xOffset + (width / 2)) + (tooltipWidth / 2);
```
So `xOffset` is always measured from the left edge.... and we are dividing the width of _something_ by `2` (why?) and then also dividing the tooltip width by `2`. Having a hard time visualizing this maybe because I'm unsure what `width` is referring to.

roryabraham replied:
Renamed `width` -> `componentWidth`. Not sure how useful such a comment would be, but what we're doing is:

1. `left edge of the wrapped component` + `half the component's width` = `center of component`
2. `center of component` - `half the tooltip's width` = `left edge of tooltip` (because the tooltip is centered on the component by default.

marcaaron replied:
I would say that is super useful. I'm kind of a visual learner and guessing others are as well so painting a bit of a picture in words seems like a good idea for some of these.

roryabraham replied:
Moved the horizontal shifting logic to its own function and I think it's easier to understand now.

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,117 @@
+        // Tooltip is in right gutter, shift left
+        horizontalShift = (windowWidth - GUTTER_WIDTH) - tooltipRightEdge;
+    }
+    horizontalShift = roundToNearestMultipleOfFour(horizontalShift);
```
Maybe naming this  `gutterOffset` would help to visualize what this value will be used for eventually. Sorta lost on how this all works.

roryabraham replied:
idk I feel like `horizonalShift` is a better name, because it's the value by which the tooltip is shifted to the left or right.

marcaaron replied:
`additionalOffsetX`?

marcaaron replied:
Alright. I personally think having a variable named `shift` doesn't read as clearly as a value because it sounds like a verb. But maybe that's just me.

marcaaron reviewed (changes requested):
Some initial comments still haven't tested just trying to wrap my head around everything so far and ask questions.

roryabraham committed:
Separate and clarify measureInWindow methods

roryabraham commented on src/components/Hoverable/index.js:
```diff
@@ -0,0 +1,57 @@
+        return React.cloneElement(React.Children.only(child), {
+            onMouseEnter: this.handleMouseEnter,
+            onMouseLeave: this.handleMouseLeave,
+        });
```
Yeah, this was mostly just there because it was taken from the implementation (from the author of `react-native-web`) that I was basing this off of. I _think_ I can remove the only child restriction from the `Hoverable` without any issues. 

However, I think that the tooltip _should_ have only one child, because otherwise the placement of the tooltip will be weird/unpredictable.

marcaaron replied:
>otherwise the placement of the tooltip will be weird/unpredictable

Why would it be weird or unpredictable? Wouldn't it just be the total height and width of all the wrapped elements?

roryabraham replied:
Okay, I updated this so `Hoverable` can have multiple children. Tooltip can too, but I left a comment warning that you shouldn't try and do that.

marcaaron replied:
Alright, and any thoughts on why we should not use a `View` to wrap the children? e.g. just add another `div` around the multiple children and we'd set the mouse events on that. Maybe I am missing it, but why should we warn anyone not to do this?

roryabraham replied:
Updated `Hoverable` to use a wrapper `View` instead of `React.Children.map` and `React.cloneElement`.

Using a wrapper `View` technically might be slightly less robust – In the case where something other than a React element (such as text outside of a `<Text>) is provided as a child to the `Hoverable`, we'll attempt to render that and probably crash native platforms. But that would happen outside of the `Hoverable` anyways, so it's probably fine to do it with the wrapper `View`, because it's easier to understand.

tgolen replied:
> In the case where something other than a React element (such as text outside of a <Text>)

I think we don't need to worry about cases never exist :D You can't have text outside of a `<Text>` in RN and you can't ever render anything that's not a React element (unless I'm missing something).

roryabraham commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,135 @@
+        // The wrapper view containing the wrapped content along with the Tooltip itself.
+        this.wrapperView = null;
+        this.width = 0;
+        this.height = 0;
```
Just updated these to be:

```
this.wrapperWidth = 0;
this.wrapperHeight = 0;
```

roryabraham committed:
Clean up horizontal shifting logic

roryabraham committed:
Merge branch 'master' into Rory-Tooltips

roryabraham committed:
Use withWindowDimensions HOC

roryabraham committed:
Make hoverable work with multiple children

roryabraham commented on src/components/Hoverable/index.js:
```diff
@@ -0,0 +1,57 @@
+            : this.props.children;
+
+        // Clone the child element, providing the callbacks to the react element to handle hover state changes.
+        // Using `React.Children.only` enforces that this component has only one child,
```
I updated this to let `Hoverable` take multiple children.

marcaaron commented on src/components/Hoverable/index.js:
```diff
@@ -0,0 +1,49 @@
+            this.props.onHoverIn();
+            this.setState({isHovered: true});
+        }
+    }
```
What does `this.state.isHovered` do exactly it seems unused? Am I wrong in thinking that we do not need a `isHovered` state and are just replacing `onMouseEnter` and `onMouseLeave` with `onHoverOut/onHoverIn`?

roryabraham replied:
`this.state.isHovered` allows `Hoverable` to have parity with the hover-state interface of `Pressable`, so it can be used like this:

```javascript
<Hoverable>
    {hovered => (
        <View style={{hovered 
             ? themeColors.hoveredComponentBG 
             : themeColors.componentBG}}
        />
    )}
</Hoverable>
```

When you made this comment, I had unknowingly broken this functionality, but it's fixed now.

marcaaron replied:
Ok haha cool I wasn't sure if we were still going with that. Sounds good to me.

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,141 @@
+     * Measure the size and position of the wrapper view.
+     */
+    measureWrapperAndGetPosition() {
+        this.wrapperView.measureInWindow((x, y, width, height) => {
```
Might be nice to mention somewhere in the code why we are using `measureInWindow`

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,156 @@
+            // = 26
+            // OR
+            // Shift the tooltip down by...
+            // (component height) + (pointer height - 1)
```
Maybe this would be clearer if this was also a variable with an explanation about how we came up with this number? If these values were to change would we need to recalculate the `26` again?

marcaaron replied:
Same goes for the `3`. Maybe would be better as a constant.

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,156 @@
+            top: shouldShowBelow ? componentHeight - 4 : -4,
+
+            // Shift the pointer to the right (to the middle of the wrapped element)
+            left: (componentWidth / 2) - 4,
```
Looks like this 4 has a relationship with the 4 in the `pointerStyle` so maybe it should use a variable.

roryabraham committed:
Extract more variables in styles calculations

roryabraham committed:
Add comment about measureInWindow

roryabraham committed:
Fix and simplify Hoverable

roryabraham commented on src/components/Hoverable/index.js:
```diff
@@ -0,0 +1,52 @@
+            ? this.props.children(this.state.isHovered)
+            : this.props.children;
+        return (
+            <View
```
Changed this to just use a wrapper `View` instead of all the `React.Children.map` and `React.cloneElement` stuff, which should be fine, so long as the children provided are valid react elements. I figure they always will be, so this is safe enough.

marcaaron reviewed (dismissed):
LGTM. Still a couple of things I think might not be necessary but nice work overall and thanks for making the changes.

flodnv commented on src/components/Hoverable/HoverablePropTypes.js:
```diff
@@ -0,0 +1,22 @@
+import PropTypes from 'prop-types';
+
+const propTypes = {
+    // Child to wrap with Hoverable.
```
Child -> Children ?

flodnv commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,145 @@
+    // The text to display in the tooltip.
+    text: PropTypes.string.isRequired,
+
+    // Child to wrap with Tooltip. Should be only child, or the tooltip placement will be unpredictable.
```
Child -> Children ? 

What does `only child` mean?

flodnv commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,176 @@
+            // Shift the tooltip down (+) by...
+            //
+            //   component height
+            // + (pointer height - 1)
```
Why `1` and not `4` ?

roryabraham replied:
By default, the tooltip top is at the top of the component. So first it's shifted down by the component height so that the tooltip top is at the component bottom. Then the pointer is in the way, but if you shift it down by the height of the pointer, `4`, then there is a small but noticeable gap (presumably 1px) between the tooltip and the pointer. If you shift it down by `3` it looks right.

flodnv reviewed (commented):
I see only Web has been tested, should other platforms be checked off just to be sure they're not totally broken?

roryabraham committed:
Fix prop comments and remove collapsable=false

roryabraham committed:
Remove unnecessary interpolation

roryabraham committed:
Fix initial load and window resize issues

roryabraham committed:
Fix weird zIndex trasparency issue

roryabraham committed:
Remove irrelevant zIndex style change

roryabraham commented:
Fixed a few more bugs I found while testing this:

- I found that the tooltips would sometimes (inconsistently) display wrong when the page first loads, but that it would always be fixed after clicking anywhere on the screen. This puzzled me at first, but after a bit I figured out that many of the class fields needed to be in state, so that a re-render occurs when they're updated.
- Also, the measurement of the wrapper view wasn't happening when the window was resized, so I moved the measurement to its own function and call it in `componentDidUpdate` if the `windowDimensions` have changed.

Luke9389 commented:
@roryabraham Great work on the PR so far! I've been super busy with other stuff, but you've had enough cooks in the kitchen without me I think :) _(four other reviewers!)_

Would you mind expounding a little bit on that first bullet point? I'm not sure I understand what you're pointing to.

roryabraham commented:
> Would you mind expounding a little bit on that first bullet point?

Yeah, check out the changes from the commit titled "Fix initial load and window resize issues". Basically the following fields were just properties of the tooltip class, but needed to be in the state: `xOffset`, `yOffset`, `wrapperWidth`, `wrapperHeight`, `tooltipWidth`, `tooltipHeight`, for two reasons:

1) They can change, and when they do those changes need to be reflected in the UI
2) When the app first loaded, `measureInWindow` was happening asynchronously, so the result wasn't always stored by the time the component rendered. Instead, I needed to call `setState` in the callback for `measureInWindow` to store the updated values and trigger a re-render.

roryabraham commented:
Anyways, I finished testing and confirmed that this works on desktop as expected, and didn't notice any regressions on the other platforms.

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,180 @@
+        // because `measureInWindow` provides the x and y offset relative to the window, rather than the parent element.
+        this.getWrapperPosition()
+            .then(({x, y}) => {
+                // Re-render component only if values have changed
```
Another option would be to return `null` in `setState` if you don't want to re-render. Sometimes can make optimizations like this cleaner. Or just re-render anyway unless there is some reason not to. It's not always valuable to prevent every re-render.

marcaaron replied:
Unless you are doing this to prevent an infinite loop... in that case, might be better handled by `shouldComponentUpdate()`

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * Measure the size of the tooltip itself.
+     */
+    measureTooltip({nativeEvent}) {
```
missing param in doc

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+ * @returns {Object}
+ */
+export default function getTooltipStyles(
+    interpolatedSize,
```
Not interpolated anymore. Maybe better to call this `scale` or `scaleSize`.

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    componentDidUpdate(prevProps) {
+        if (!_.isMatch(this.props.windowDimensions, prevProps.windowDimensions)) {
```
I think maybe this should be `isEqual`? Not really sure if `isMatch` does the same thing https://underscorejs.org/#isMatch

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,180 @@
+    /**
+     * Measure the size and position of the wrapper view.
+     */
+    measureWrapperAndGetPosition({nativeEvent}) {
```
missing param in doc

roryabraham committed:
cleanup and use shouldComponentUpdate

roryabraham committed:
Merge branch 'master' into Rory-Tooltips

# Conflicts:
#	src/styles/styles.js

marcaaron commented on src/components/Hoverable/index.js:
```diff
@@ -0,0 +1,52 @@
+
+    render() {
+        // If this.props.children is a function, call it to provide the hover state to the children.
+        const childrenWithHoverState = _.isFunction(this.props.children)
```
pretty small NAB, but we don't really need this variable `childrenWithHoverState` and could just use `{}`

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,183 @@
+        if (!_.isEqual(this.props.windowDimensions, prevProps.windowDimensions)) {
+            this.getWrapperPosition()
+                .then(({x, y}) => {
+                    if (x !== this.state.xOffset || y !== this.state.yOffset) {
```
I think this check is no longer needed. We won't re-render if these state value have not changed so we should be good.

marcaaron commented on src/components/Tooltip.js:
```diff
@@ -0,0 +1,183 @@
+           We only want to re-render if props or state have changed.
+           This prevents an infinite rendering loop with the `onLayout` callback calling setState.
+           Also, we need to do a deep comparison of props using _.isEqual rather than relying on the shallow comparison
+           done by `PureComponent` because a shallow comparison would not catch changes in `props.windowDimensions`.
```
NAB. 

This makes sense, but it does have me wondering about why the infinite rendering loop is happening and whether there is no better way to prevent it.

Some questions I have...
- In what scenario do we need to re-measure the items? 
- What exactly causes a "layout" to occur?
- Is `onLayout` typically called on every `render()`?

Separately, we could maybe get a bit more specific about the conditions where we want to allow a render (e.g. "only re-render when `(this.state.xOffset !== nextState.xOffset) || etc`").

Not sure we need to block on this per se, but I am curious to learn more about what is happening with `onLayout`.

roryabraham replied:
> In what scenario do we need to re-measure the items

It's difficult to know when we need to re-measure the wrapper view, because it could be anything. It could be completely static, or it could be constantly changing – we don't really know what the children are doing. At minimum, we need to measure the wrapper view initially and when the window dimensions change.

The tooltip would only need to be re-measured if `this.props.text` changes.

> What exactly causes a "layout" to occur

I'm really not sure ... I couldn't find the answer to this question in the docs.

> Is onLayout typically called on every render()

I thought so, but maybe not ... according to the docs, `onLayout` is "Invoked on mount and layout changes".

> Separately, we could maybe get a bit more specific about the conditions where we want to allow a render

I originally had this fully-expanded like you're saying, but I think we would want a re-render to occur if any state or prop changes:

[ ] - `this.props.text` - Yes because this would change the tooltip width and height
[ ] - `this.props.children` - Yes because this would change the wrapper width, height, maybe position too
[ ] - `this.props.windowDimensions` - Yes because this would make `this.state.xOffset` and `this.state.yOffset` incorrect.
[ ] - `this.state.xOffset` and `this.state.yOffset` - Yes because otherwise the tooltip will display offset from the component.
[ ] - `this.state.XXXwidth` and `this.state.XXXheight` - Yes because the updated width and height (resulting from the change in `this.props.text` or `this.props.children` is not available until after the next `render` -> `onLayout` -> `render`

marcaaron replied:
Thanks for exploring that. Cool with this for now, but still curious if there really is any "infinite rendering" problem here (which would seem to indicate that `onLayout` is getting called every render) and if so whether there's anyway around it. 

Another future improvement here might be to refactor `withWindowDimensions` so it returns less complex props (something like `width`, `height`, and `isSmallScreen` instead of an object). Then we can use `PureComponent` or `React.memo()` here to solve this problem.

roryabraham replied:
Cool, NAB for now so I'm going to resolve

marcaaron commented on src/styles/utilities/spacing.js:
```diff
@@ -117,6 +117,16 @@ export default {
 
+    pv1: {
+        paddingTop: 4,
+        paddingBottom: 4,
```
Can we use `paddingVertical`?

marcaaron commented on src/styles/utilities/spacing.js:
```diff
@@ -117,6 +117,16 @@ export default {
+
+    ph2: {
+        paddingRight: 8,
+        paddingLeft: 8,
```
`paddingHorizontal`

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+            ...tooltipVerticalPadding,
+            ...spacing.ph2,
+
+            // Shift the tooltip down (+) by...
```
What does `(+)` mean here?

marcaaron replied:
"a positive number of pixels" ?

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+            // + bottom padding
+            // + upward shift of the pointer (pointer height)
+            // + (pointer height - 1)
+            //
```
I think this comments could be more valuable if they did more than just repeat what we are doing in the code below and would maybe work better as a short sentence that reminds us why we are doing this.

>If we are showing the tooltip below we set the `top` value to be equal to the component height plus the height of the pointer minus 1 extra pixel for _____ to push things below the wrapped content.

>If we are showing the tooltip above then we must apply a negative `top` value and also take into account the size of the font, overall vertical padding and multiply the pointer height by 2 because...

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+            // + (pointer height - 1)
+            //
+            top: shouldShowBelow
+                ? componentHeight + (POINTER_HEIGHT - 1)
```
What is the `- 1` for?

marcaaron replied:
(comment in the code maybe or give a relevant variable name so it's clearer)

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+                    tooltipFontSize
+                    + tooltipVerticalPadding.paddingTop
+                    + tooltipVerticalPadding.paddingBottom
+                    + ((2 * POINTER_HEIGHT) - 1)
```
Why do we multiply the pointer height by 2 ?

marcaaron replied:
Also is the `- 1` the same as the one above?

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+            //
+            top: shouldShowBelow
+                ? componentHeight + (POINTER_HEIGHT - 1)
+                : -1 * (
```
I _think_ this would also work for getting a negative number...

```suggestion
                : -(
```

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+            //   half the component's width
+            // + half the tooltip's width
+            // + any horizontal shift
+            left: (tooltipWidth / -2) + (componentWidth / 2) + horizontalShift,
```
Might be clearer to just make the first value negative rather than divide by a negative number. 

It's slightly unclear why the first value is negative? Maybe we can add a sentence to explain where the tooltip will end up positioned based on this instead of the summary of the code in the above comment?

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,177 @@
+            top: shouldShowBelow ? (componentHeight - POINTER_HEIGHT) : (-1 * POINTER_HEIGHT),
+
+            // Shift the pointer to the right (to the middle of the wrapped element)
+            left: (componentWidth / 2) - (POINTER_WIDTH / 2),
```
Pretty much similar comments as before. Would be nice to add brief context for what the visual end result will be, why we are doing things, and how the math supports it etc.

marcaaron reviewed (changes requested):
This is looking really good. I'd love to see the styling section get cleaned up a bit if possible. To be honest, I'm having trouble connecting some of the logic in the styles with how things will look visually and some of the math could use a little more explanation.

roryabraham committed:
Remove unnecessary local variable from hoverable

roryabraham committed:
Get rid of unnecessary state change check in componentDidUpdate

roryabraham committed:
Use paddingHorizontal and paddingVertical

roryabraham committed:
Cleanup tooltip styles and make pointer wider

roryabraham committed:
Add manual shift props as final touch

roryabraham commented:
I cleaned up the style a fair amount and added a ton of more detailed comments so hopefully things are a bit clearer now. Also I added a couple props `manualShiftHorizontal` and `manualShiftVertical` so that you can pass those in to make finishing touches to the tooltip placement if necessary. Some examples of where that comes in handy:

1) A tooltip will be covered by another element but is not close to the edge of the screen:

|Before|After|
|---|---|
|![image](https://user-images.githubusercontent.com/47436092/106402181-3733dc80-63dd-11eb-8782-4fb996b295f1.png)|![image](https://user-images.githubusercontent.com/47436092/106402110-d906f980-63dc-11eb-8229-df929060d2ad.png)|

2) A child component of the tooltip has some vertical margin, but we want the tooltip pointer to point directly at the child

|Before|After|
|---|---|
|![image](https://user-images.githubusercontent.com/47436092/106402161-1f5c5880-63dd-11eb-9979-e5bba025aa52.png)|![image](https://user-images.githubusercontent.com/47436092/106402140-02278a00-63dd-11eb-83d4-5b4db65650e9.png)|

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,198 @@
+
+            // Because it uses absolute positioning, the top-left corner of the tooltip is aligned
+            // with the top-left corner of the wrapped component by default.
+            // So we need to shift the tooltip vertically and horizontally to position it correctly.
```
❤️

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,198 @@
+            //      so that the top of the pointer aligns with the bottom of the component.
+            //
+            // Always add the manual vertical shift passed in as a parameter.
+            top: shouldShowBelow ? componentHeight + manualShiftVertical : manualShiftVertical - POINTER_HEIGHT,
```
Is `manualShiftVertical` a required parameter? If not, should this have some kind of default value?

marcaaron commented on src/styles/getTooltipStyles.js:
```diff
@@ -0,0 +1,198 @@
+            //      so the tooltip's center lines up with the center of the wrapped component.
+            //   3) Add the horizontal shift (left or right) computed above to keep it out of the gutters.
+            //   4) Lastly, add the manual horizontal shift passed in as a parameter.
+            left: ((componentWidth / 2) - (tooltipWidth / 2)) + horizontalShift + manualShiftHorizontal,
```
Same question about required parameter `manualShiftHorizontal`. Are we going to always need to pass these props?

marcaaron reviewed (dismissed):
Looks much better thanks for all the docs. Had a few more comments/questions but they are nabs and can wait for the implementation since I'm probably missing something.

roryabraham committed:
Make manualShiftHorizontal and manualShiftVertical optional in getTooltipStyles

roryabraham commented:
I made both `manualShiftHorizontal` and `manualShiftVertical` optional and gave them both a default value of 0.

flodnv commented:
NAR -- leaving merge to whoever wants it (maybe @tgolen ?)
